doc/ni_by_example_4.md

```diff
@@ -1,169 +1,763 @@
+I don't like the of Perl's `substr` method because it's asymmetric. to recover most of what I like about from Python's string slicing syntax, I use regex.
 
-##Matrix Operations
+```
+$ ni iabcdefgh p'/^(.*).{4}$/'  #[:-4] in Python
+abcd
+(END)
+```
 
-Operations on huge matrices are not entirely `ni`ic, since they may require space greater than memory, whichwill make them slow. However, operators are provided to improve These operations are suited best to 
+```
+$ ni iabcdefgh p'/^.{3}(.*)$/' #[3:] in Python
+defgh
+(END)
+```
 
+```
+$ ni iabcdefgh p'/^.*(.{2})$/' #[-2:] in Python
+gh
+(END)
+```
 
-* `X<col>`, `Y<col>`, `N<col>`: Matrix Partitioning
-  * **TODO**: understand how these actually work.
-* `X`: sparse-to-dense transformation
-  * In the case that there are collisions for locations `X`, `X` will sum the values
-  * For example: `ni n010p'r 0, a%3, 1' X`
+####Using Capture Groups
 
-##Disk-Backed Data Closures
+Capture groups are set off using parentheses; to get them explicitly,  syntactic sugar for the more explicit `my @v = $_ =~ /regex/;`
 
-* `@:[disk_backed_data_closure]`
+This will get capture groups and store them in `@v`. Using parentheses around `$v` tells perl to use it in a list context, which will allow you to capture one or more groups.
 
-##Annoyingly Advanced Perl
-* `use strict` and the `::` prefix in a Perl Environment
-  * When `use strict` is enabled, Perl will complain when you try to create a variable in a Perl snippet that does not start with `::`.
-  * The reasons for this are very specific to Perl; if you are a true Perl nerd, you can look them up, but you do not need to know them if you just accept that variables need to start with `::` when you apply `use strict`.
-  * It is probably a good idea to `use strict` when the variables you define are sufficiently complex; otherwise you're probably okay not using it.
-  
-##Binary Operations
-In theory, this can save you a lot of space. But I haven't used this in practice.
-  
+For example:
+
+```
+$ ni iabcdefgh p'my @v = /^(.)(.)/; r @v'
+a       b
+(END)
+```
+
+```
+$ ni iabcdefgh p'my ($w) = /^(.)/; r $w'
+a
+(END)
+```
+
+```
+$ ni iabcdefgh p'my ($x, $y) = /^(.)(.)/; r $x, $y'
+a       b
+(END)
+```
+
+
+####Substitution `s///`, Translation `tr///` and `y///`
+
+These operators have a slightly tricky syntax. For example, you can't use these operators the way you'd use capture groups. 
+
+```
+$ ni iabcdefgh p'tr/a-z/A-Z/'
+8
+(END)
+```
 
-##Less Useful `ni`-specific Perl Extensions
+```
+$ ni iabcdefgh p's/abc/ABC/'
+1
+(END)
+```
 
+The reason for these somewhat surp The return value of `tr` and `y` is the number of characters that were translated, and the return value of `s` is 0 if no characters were substituted and ` if characters were.
+This also will give somewhat-surprising behavior to code like:
 
-###Array Functions
-  * `clip`
-  * `within`
+```
+$ ni iabcdefgh p'$v = tr/a-z/A-Z/; $v'
+8
+```
+
+Instead, these operators work as side-effects.
```

Comment:
There's a variant available in perl 5.12 or 5.14 onwards: `s/foo/bar/r`: the `r` indicates that we want to return the modification rather than applying it. But this will fail if ni is running on a system with a very old perl.